src/main/java/com/salesforce/dynamodbv2/mt/mappers/sharedtable/impl/RandomPartitioningConditionMapper.java

```diff
@@ -30,143 +38,272 @@
-    void convertFieldNameLiteralsToExpressionNames(RequestWrapper request, Collection<FieldMapping> fieldMappings) {
-        request.setPrimaryExpression(
-            convertFieldNameLiteralsToExpressionNamesInternal(fieldMappings, request.getPrimaryExpression(), request));
-        request.setFilterExpression(
-            convertFieldNameLiteralsToExpressionNamesInternal(fieldMappings, request.getFilterExpression(), request));
+    @Override
+    public void applyForUpdate(RequestWrapper request) {
+        String updateExpression = request.getPrimaryExpression();
+        String conditionExpression = request.getFilterExpression();
+        Set<String> virtualFieldNames = tableMapping.getAllMappingsPerField().keySet();
+
+        updateExpression = convertFieldNameLiteralsToExpressionNames(updateExpression, request, virtualFieldNames);
+        validateNotUpdatingTablePrimaryKeyFields(updateExpression, request);
+
+        if (conditionExpression != null) {
+            conditionExpression = convertFieldNameLiteralsToExpressionNames(conditionExpression, request,
+                virtualFieldNames);
+            conditionExpression = makePlaceholdersDistinct(updateExpression, conditionExpression, request);
+            request.setFilterExpression(conditionExpression);
+            applyToFilterExpression(request, false);
+        }
+
+        for (Entry<String, List<FieldMapping>> entry : tableMapping.getAllMappingsPerField().entrySet()) {
+            updateExpression = mapFieldInUpdateExpression(updateExpression, request, entry.getKey(), entry.getValue());
+        }
+        request.setPrimaryExpression(updateExpression);
     }
 
     @Override
-    public void apply(RequestWrapper request) {
-        apply(request, null);
+    public void applyToKeyCondition(RequestWrapper request, @Nullable DynamoSecondaryIndex virtualSecondaryIndex) {
+        Collection<FieldMapping> fieldMappings = virtualSecondaryIndex == null
+            ? tableMapping.getTablePrimaryKeyFieldMappings()
+            : tableMapping.getIndexPrimaryKeyFieldMappings(virtualSecondaryIndex);
+
+        String expression = request.getPrimaryExpression();
+        expression = mapFieldsInConditionExpression(expression, request, fieldMappings);
+        request.setPrimaryExpression(expression);
     }
 
     @Override
-    public void apply(RequestWrapper request, @Nullable DynamoSecondaryIndex virtualSecondaryIndex) {
-        Collection<FieldMapping> fieldMappings;
-        if (virtualSecondaryIndex == null) {
-            fieldMappings = tableMapping.getTablePrimaryKeyFieldMappings();
+    public void applyToFilterExpression(RequestWrapper request, boolean isPrimaryExpression) {
+        String expression = isPrimaryExpression ? request.getPrimaryExpression() : request.getFilterExpression();
+
+        // for conditional / filter expressions, for each indexed field, we only need to use one of its field mappings,
+        // since we're supporting only equals expressions
```

Comment:
This comment is outdated, before I added support for non-equals comparison operators when using non-legacy filter expressions... I'm not sure there's a bug. I'm confused because I'd expect there to be a bug if we do >, >=, <, <= comparisons on a number RK, when using binary encoding, because I thought we didn't have a binary encoding for numbers that preserves order yet...

Reply:
On okay I was confused because I was thinking currently mt-dynamo already supports mapping a virtual range key of type N to a physical range key of type B. But that's not true and the range key types have to be exactly the same right now. As long as we only allow RK mappings where the order is preserved, we can use any of the index FieldMappings to map comparison clauses for a given virtual field.